lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart

```diff
@@ -101,5 +110,49 @@ class FlutterViewManager {
     // close the streams.
     _onViewCreatedController.close();
     _onViewDisposedController.close();
+    // Remove the focus tracking event listeners.
+    domDocument.body?.removeEventListener(_focusin, _focusChangeHandler, true);
+    domDocument.body?.removeEventListener(_focusout, _focusChangeHandler, true);
+  }
+
+  int? _latestFocusedViewId;
+
+  // TODO(tugorez): Add handlers for backwards focus.
+  late final DomEventListener _focusChangeHandler = createDomEventListener((DomEvent event) {
+    final int? viewId = _findParentViewIdOrNull(domDocument.activeElement);
+    if (viewId == _latestFocusedViewId) {
+      return;
+    }
+
+    late final ui.ViewFocusEvent event;
+    if(viewId == null) {
+       event = ui.ViewFocusEvent(
+         viewId: _latestFocusedViewId!,
+         state: ui.ViewFocusState.unfocused,
+         direction: ui.ViewFocusDirection.undefined,
+       );
+      _latestFocusedViewId = null;
+    } else {
+      event = ui.ViewFocusEvent(
+        viewId: viewId,
+        state: ui.ViewFocusState.focused,
+        direction: ui.ViewFocusDirection.forward,
+      );
+      _latestFocusedViewId = viewId;
+    }
+    _dispatcher.invokeOnViewFocusChange(event);
+  });
+
+  static int? _findParentViewIdOrNull(DomElement? element) {
+    if (element == null) {
+      return null;
+    }
+    if (element.tagName.toLowerCase() == DomManager.flutterViewTagName) {
+      return int.parse(element.getAttribute(GlobalHtmlAttributes.flutterViewIdAttributeName) ?? '0');
+    }
+    return _findParentViewIdOrNull(element.parent);
```

Comment:
Please, use the `closest()` method. [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest). Also `int.tryParse` so this method doesn't throw. [Dart]().

```dart
return int.tryParse(
  element.closest(DomManager.flutterViewTagName)
    ?.getAttribute(GlobalHtmlAttributes.flutterViewIdAttributeName) ?? 'null',
  radix: 10,
);
```

Also also, if you return `int?` I don't think you need to add `OrNull` to the name of the method, just add a little bit of documentation that says "returns the viewId of the parent flutter view of this element, or null if the element is not contained within a flutter view." or similar.

Also also also, in the default, you're returning '0', which may be a valid `viewId`. I'd expect this method to return null when a parent view is not found.

Reply:
Also, instead of reading the `flt-view-id` from the DOM, maybe we want to search for the EngineFlutterView that is holding on to the DOM Element and retrieve its viewId from the ViewManager directly, so we don't rely on the DOM as data? @mdebbar? (For now, I think this should be enough though!)

Reply:
> so we don't rely on the DOM as data?

In order to do this, we can either have a separate `focusin/focusout` listener for each view, or maintain a Map from `rootElement` => `FlutterView`. I need something similar in my coming PR for multi-view text editing. I'll try to figure something out in that PR.

Reply:
SGTM, happy to refactor this in a follow up PR.